presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -0,0 +1,1970 @@
+        return delegate.getDatabase(databaseName);
+    }
+
+    public synchronized Optional<List<String>> getAllTables(String databaseName)
+    {
+        checkReadable();
+        if (!tableActions.isEmpty()) {
+            throw new UnsupportedOperationException("Listing all tables after adding/dropping/altering tables/views in a transaction is not supported");
+        }
+        return delegate.getAllTables(databaseName);
+    }
+
+    public synchronized Optional<Table> getTable(String databaseName, String tableName)
+    {
+        checkReadable();
+        Action<TableAndMore> tableAction = tableActions.get(new SchemaTableName(databaseName, tableName));
+        if (tableAction == null) {
+            return delegate.getTable(databaseName, tableName);
+        }
+        else {
+            switch (tableAction.getType()) {
+                case ADD:
+                case ALTER:
+                case INSERT_EXISTING:
+                    return Optional.of(tableAction.getData().getTable());
+                case DROP:
+                    return Optional.empty();
+                default:
+                    throw new IllegalStateException("Unknown action type");
+            }
+        }
+    }
+
+    /**
+     * This method can only be called when the table is known to exist
+     */
+    private TableSource getTableSource(String databaseName, String tableName)
+    {
+        checkReadable();
+        Action<TableAndMore> tableAction = tableActions.get(new SchemaTableName(databaseName, tableName));
+        if (tableAction == null) {
+            return TableSource.PRE_EXISTING_TABLE;
+        }
+        switch (tableAction.getType()) {
+            case ADD:
+                return TableSource.CREATED_IN_THIS_TRANSACTION;
+            case ALTER:
+                throw new IllegalStateException("Tables are never altered in the current implementation");
+            case DROP:
+                throw new TableNotFoundException(new SchemaTableName(databaseName, tableName));
+            case INSERT_EXISTING:
+                return TableSource.PRE_EXISTING_TABLE;
+            default:
+                throw new IllegalStateException("Unknown action type");
+        }
+    }
+
+    public synchronized Optional<List<String>> getAllViews(String databaseName)
+    {
+        checkReadable();
+        if (!tableActions.isEmpty()) {
+            throw new UnsupportedOperationException("Listing all tables after adding/dropping/altering tables/views in a transaction is not supported");
+        }
+        return delegate.getAllViews(databaseName);
+    }
+
+    /**
+     * {@code currentLocation} needs to be supplied if a writePath exists for the table.
+     */
+    public synchronized void createTable(ConnectorSession session, Table table, PrincipalPrivilegeSet principalPrivilegeSet, Optional<Path> currentPath)
+    {
+        setShared();
+        checkNoPartitionAction(table.getDatabaseName(), table.getTableName());
```

Comment:
I don't think this is needed, since you should not able to create a table unless the table currently does not exist and in that cause you could not have partition actions.